src/_pytest/outcomes.py

```diff
@@ -141,21 +150,62 @@ def skip(msg: str = "", *, allow_module_level: bool = False) -> "NoReturn":
         to skip a doctest statically.
     """
     __tracebackhide__ = True
-    raise Skipped(msg=msg, allow_module_level=allow_module_level)
+    reason = _resolve_msg_to_reason("skip", reason, msg)
+    raise Skipped(msg=reason, allow_module_level=allow_module_level)
 
 
 @_with_exception(Failed)
-def fail(msg: str = "", pytrace: bool = True) -> "NoReturn":
+def fail(
+    reason: str = "", pytrace: bool = True, msg: Optional[str] = None
+) -> "NoReturn":
     """Explicitly fail an executing test with the given message.
 
-    :param str msg:
+    :param str reason:
         The message to show the user as reason for the failure.
+
     :param bool pytrace:
         If False, msg represents the full failure information and no
         python traceback will be reported.
+
+    :param str msg:
+        Temporary string to support deprecating msg=, will be removed later.
+    """
+    __tracebackhide__ = True
+    reason = _resolve_msg_to_reason("fail", reason, msg)
+    raise Failed(msg=reason, pytrace=pytrace)
+
+
+def _resolve_msg_to_reason(
+    func_name: str, reason: str, msg: Optional[str] = None
+) -> str:
+    """
+    Handles converting the deprecated msg parameter if provided into
+    reason, raising a deprecation warning.  This function will be removed
+    when the optional msg argument is removed from here in future.
+
+    :param str func_name:
+        The name of the offending function, this is formatted into the deprecation message.
+
+    :param str reason:
+        The reason= passed into either pytest.fail() or pytest.skip()
+
+    :param str msg:
+        The msg= passed into either pytest.fail() or pytest.skip().  This will
+        be converted into reason if it is provided to allow pytest.skip(msg=) or
+        pytest.fail(msg=) to continue working in the interim period.
+
+    :returns:
+        The value to use as reason.
+
     """
     __tracebackhide__ = True
-    raise Failed(msg=msg, pytrace=pytrace)
+    if msg is not None:
+        from _pytest.deprecated import KEYWORD_MSG_ARG
+
+        assert reason == ""
```

Comment:
not sure if this adds any value

Reply:
I think it does, because we don't want the user to pass both `reason` and `msg`, so it is correct to error out here.

However I think we should use a proper exception with a clear message:

```suggestion
        if reason:
            fail(f"Cannot pass both ``reason`` and ``msg`` arguments to {func_name}.")
```